### DISCUSSION TITLE
Can I use react-native-web with Vite instead of WebPack?

### DISCUSSION THREAD

**Discussion by hxse** [necolas/react-native-web#2201] [Q&A] [answered]
Vite is very fast.
Have any demos?

**Comment by hxse** [comment]
I found a great example in youtube:
https://www.youtube.com/watch?v=2wOvhDtqsW8

**Reply by sagarkudutata** [reply]
thanks

**Comment by hxse** [comment, marked as answer]
I found a way:
* `npx react-native init demo`
* `cd demo`
* Add following code in `metro.config.js` -> `module.exports = {}`
  This is to compatibility with JSX files.
  Maybe you need to reinstall the APP will take effect.
```
resolver: {
sourceExts: ['js', 'jsx', 'json', 'ts', 'tsx', 'mjs'],
  }
```
* `yarn create vite web --template react`
* `cd web && yarn`
* Add following code in `vite.config.js` ->  `export default defineConfig({})`
```
resolve: {
    alias: {
      'react-native': 'react-native-web',
    },
  },
```
* `mkdir share && touch share/test.jsx`
  Add following code in `share/test.jsx`
```
import React from 'react';
import {Text} from 'react-native';

const Test = ({title}) => {
  return <Text>hello world { title}</Text>;
};

export default Test;
```
* `import Section from './web/src/share/test';` // import in RN-dir App.js and add in App component.
* `import Section from './share/test';` // import in Web-dir App.jsx and add in App component.
* `yarn dev`
* `cd .. && yarn start`

**Reply by dalcib** [reply]
See https://github.com/vitejs/vite/discussions/8195#discussioncomment-3053838

**Reply by Marknjo** [reply]
Almost an year from when this was posted, and it has been super helpful. @hxse, Thank you. Incase anyone out there comes across the same problem, here is an example setup with Vite an react-native-web. As am writing this I'm using storybooks v7 

```ts
import type { StorybookConfig } from '@storybook/react-vite'

import { join, dirname, resolve } from 'path'

/**
 * This function is used to resolve the absolute path of a package.
 * It is needed in projects that use Yarn PnP or are set up within a monorepo.
 */
function getAbsolutePath(value: string): any {
  return dirname(require.resolve(join(value, 'package.json')))
}

const config: StorybookConfig = {
  stories: ['../stories/**/*.stories.tsx'],
  addons: [
    getAbsolutePath('@storybook/addon-links'),
    getAbsolutePath('@storybook/addon-essentials'),
    getAbsolutePath('@storybook/addon-a11y'),
    getAbsolutePath('@storybook/addon-styling'),
    getAbsolutePath('@storybook/addon-onboarding'),
    getAbsolutePath('@storybook/addon-interactions'),
    getAbsolutePath('@storybook/addon-react-native-web'),
  ],
  framework: {
    name: '@storybook/react-vite',
    options: {},
  },
  async viteFinal(config, { configType }) {
    // customize the Vite config here
    return {
      ...config,
      esbuild: {
        logOverride: {
          'this-is-undefined-in-esm': 'silent',
        },
      },
      define: {
        'process.env': {},
      },
      resolve: {
        alias: [
          {
            find: '@mj/ui',
            replacement: resolve(__dirname, '../../ui/'),
          },
          {
            find: '@comps',
            replacement: resolve(__dirname, '../components'),
          },
          // **@NOTE**: This is the most important piece, swapping **react-native** with **react-native-web**
          {
            find: 'react-native',
            replacement: 'react-native-web',
          },
        ],
      },
    }
  },
  docs: {
    autodocs: true,
  },
}

export default config

console.log({ resolve: resolve(__dirname, './') })

```

**Reply by Rossella-Mascia-Neosyn** [reply]
this file and put in the web or mobile part? @Marknjo

**Reply by Marknjo** [reply]
Hi @Rossella-Mascia-Neosyn, this is your main entry file (.storybook/main.ts), if using typescript, otherwise remove types if using .js.  

Hey, if you are still finding yourself in the weeds, I have a monorepo setup [Here](https://github.com/Marknjo/storybook-with-solito). Explore it or just bootstrap your project based on it  with **Use This Template** button

**Reply by Marknjo** [reply]
BTW you don't have to wrap your addons with getAbsolutePath(), later on it throws some error related with fs.

**Reply by stereobooster** [reply]
In case somebody looking for another solution https://stereobooster.com/posts/react-native-web-with-vite/#final-config

**Comment by Bram-dc** [comment]
I have created a Vite plugin that hopefully covers most cases: `vite-plugin-react-native-web` ([NPM](https://www.npmjs.com/package/vite-plugin-react-native-web)), Feel free to contribute to the project for your own react-native problems you encounter while using Vite with react-native-web.

**Reply by sagarkudutata** [reply]
thanks

**Comment by Nickersoft** [comment]
@Bram-dc (or anyone else in this thread): were you able to get `react-native-web` working in terms of transitive dependencies? I'm trying to use my team's UI library in Vite, where the UI lib uses RN + RN Web seeing the product is a mobile app.

However, Vite seems to only apply aliases to your project's immediate dependencies, meaning that if the UI library references "react-native", Vite isn't aliasing it to "react-native-web". Even once I install `react-native`, Vite still complains about seemingly non-ESM RN transitive dependencies like `react-native-skia`:

```
import { Skia, useFont, RoundedRect, Circle as Circle$1, Shadow, Path as Path$1, vec, Line as Line$1, DashPathEffect, useFonts, Paragraph } from '@shopify/react-native-skia';
                                     ^^^^^^
SyntaxError: Named export 'Circle' not found. The requested module '@shopify/react-native-skia' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '@shopify/react-native-skia';
const { Skia, useFont, RoundedRect, Circle: Circle$1, Shadow, Path: Path$1, vec, Line: Line$1, DashPathEffect, useFonts, Paragraph } = pkg;
```

However, these dependencies do clearly define a `module` field in their `package.json`.

The UI library's Storybook installation, that uses Webpack, loads everything fine. 

Any ideas? 🥺
